src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1333,9 +1394,49 @@ public void InvalidateHttp2Connection(Http2Connection connection)
         {
             lock (SyncObj)
             {
-                if (_http2Connection == connection)
+                Http2Connection[]? localHttp2Connections = (Http2Connection[]?) _http2Connections;
+
+                if (localHttp2Connections == null)
                 {
-                    _http2Connection = null;
+                    return;
+                }
+
+                if (localHttp2Connections.Length == 1)
+                {
+                    // Fast shortcut for the most common case.
+                    if (localHttp2Connections[0] == connection)
+                    {
+                        _http2Connections = null;
+                    }
+                    return;
+                }
+
+                for (int i = 0; i < localHttp2Connections.Length; i++)
+                {
+                    if (localHttp2Connections[i] == connection)
+                    {
+                        if (localHttp2Connections.Length > 1)
+                        {
+                            Http2Connection[] newHttp2Connections = new Http2Connection[localHttp2Connections.Length - 1];
+
+                            if (i > 0)
```

Comment:
Can this whole for-loop be simplified to:
```
int j = 0;
for (int i = 0; i < localHttp2Connections.Length; i++)
{
  // Copy all items one-by-one except for `connection`.
  if (localHttp2Connections[i] != connection)
  {
     newHttp2Connections[j++] = localHttp2Connections[i];
  }
}
```
?

Reply:
It would be slower AFAIK. I intentionally chose to use Array.Copy since it's faster than one by one copying.

Reply:
Realistically how big will the array get, how often will it change, and how much faster is the more complicated code?

Reply:
Code was simplified to just Array.IndexOf and 2 Array.Copy without a loop. (see also the [comment above](https://github.com/dotnet/runtime/pull/38748#discussion_r450199982)).